### PR TITLE
Feature: Item grouping

### PR DESCRIPTION
**Pull request by Promises**
Add a new function to deepmerge objects (this should be checked by somebody, might have issues, though i hope not)

some groups added to adamant-armour.json, black-armour.json, bronze-armour.json, bronze-weapons.json.

new command to quickly test groups, usage:
selectgroups groupA,groupB (comma seperated list of groupnames, will return all items in all groups)
filtergroups groupA,groupB (comma seperated list of groupnames, will return only items which are shared between groups)

### CODE REVIEW

**Comment by Promises** [comment]
image

File: src/engine/config/config-handler.ts [hunk 1]
```diff
@@ -70,6 +71,57 @@ export async function loadGameConfigurations(): Promise<void> {
 }
 
 
+/**
+ * find all items in all select groups
+ * @param groupKey string or array of string of which to find items connected with
+ * @return itemsKeys array of itemkeys in all select groups
+ */
+export const findItemTagsInGroup = (groupKey: string | string[]): string[] => {
```
Comment (SchauweM) [inline src/engine/config/config-handler.ts]: I'd use the plural version just so it's clear we're talking about an array of multiple entries. Same goes for all the rest in this file

**Review by SchauweM — CHANGES_REQUESTED** [review]
I like what you've done.  I don't know if there is another way to handle this tho so I'm curious what the other contributors think of your solution.

File: src/engine/config/config-handler.ts [hunk 2]
```diff
@@ -70,6 +71,57 @@ export async function loadGameConfigurations(): Promise<void> {
...
+    if(!groupKey) {
+        return [];
+    }
+
+    if(Array.isArray(groupKey)) {
+        const collection: Record<string, boolean> = {}
```
Comment (Katilith) [inline src/engine/config/config-handler.ts]: Missing a few semi-colons in this'n

File: src/engine/config/config-handler.ts [hunk 3]
```diff
@@ -70,6 +71,57 @@ export async function loadGameConfigurations(): Promise<void> {
...
+/**
+ * find all items which are shared by all the groups, and discard items not in all groups
+ * @param groupKeys groups keys which to find items shared by
+ * @return itemKeys of items shared by all groups
+ */
+export const findItemTagsInGroupFilter = (groupKeys: string[]): string[] => {
+    if(!groupKeys || groupKeys.length === 0) {
+        return [];
+    }
+    let collection: Record<string, boolean> | undefined = undefined
```
Comment (Katilith) [inline src/engine/config/config-handler.ts]: Some more semi-colons

File: src/engine/util/objects.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/**
+ * Merge two objects or arrays, first object takes priority in case where the values cannot be merged
+ * @param objectA
+ * @param objectB
+ * @return objectC combination of objectA and objectB
+ */
+export function deepMerge<T>(objectA: T, objectB: T): T {
```
Comment (Katilith) [inline src/engine/util/objects.ts:7]: Good candidate for the `common` library in the future :)

**Review by Katilith — CHANGES_REQUESTED** [review]
Only some small lint-related things, looks awesome!!

**Review by Katilith — APPROVED** [review]

**Review by SchauweM — APPROVED** [review]

**Pull request merged**